src/sage/rings/function_field/function_field_valuation.py

```diff
@@ -0,0 +1,1212 @@
+        r"""
+        Create a unique key which identifies the valuation given by ``prime``
+        on ``domain``.
+
+        TESTS:
+
+        We specify a valuation on a function field by two different means and
+        get the same object::
+
+            sage: K.<x> = FunctionField(QQ)
+            sage: v = K.valuation(x - 1) # indirect doctest
+
+            sage: R.<x> = QQ[]
+            sage: w = GaussValuation(R, valuations.TrivialValuation(QQ)).augmentation(x - 1, 1)
+            sage: K.valuation(w) is v
+            True
+
+        The normalization is, however, not smart enough, to unwrap
+        substitutions that turn out to be trivial::
+        
+            sage: w = GaussValuation(R, QQ.valuation(2))
+            sage: w = K.valuation(w)
+            sage: w is K.valuation((w, K.hom([~K.gen()]), K.hom([~K.gen()])))
+            False
+
+        """
+        from sage.categories.function_fields import FunctionFields
+        if domain not in FunctionFields():
+            raise ValueError("Domain must be a function field.")
+
+        if isinstance(prime, tuple):
+            if len(prime) == 3:
+                # prime is a triple of a valuation on another function field with
+                # isomorphism information
+                return self.create_key_and_extra_args_from_valuation_on_isomorphic_field(domain, prime[0], prime[1], prime[2])
+
+        from sage.rings.valuation.valuation_space import DiscretePseudoValuationSpace
+        if prime.parent() is DiscretePseudoValuationSpace(domain):
+            # prime is already a valuation of the requested domain
+            # if we returned (domain, prime), we would break caching
+            # because this element has been created from a different key
+            # Instead, we return the key that was used to create prime
+            # so the caller gets back a correctly cached version of prime
+            if not hasattr(prime, "_factory_data"):
+               raise NotImplementedError("Valuations on function fields must be unique and come out of the FunctionFieldValuation factory but %r has been created by other means"%(prime,))
+            return prime._factory_data[2], {}
+
+        if prime in domain:
+            # prime defines a place
+            return self.create_key_and_extra_args_from_place(domain, prime)
+        if prime.parent() is DiscretePseudoValuationSpace(domain._ring):
+            # prime is a discrete (pseudo-)valuation on the polynomial ring
+            # that the domain is constructed from
+            return self.create_key_and_extra_args_from_valuation(domain, prime)
+        if domain.base_field() is not domain:
+            # prime might define a valuation on a subring of domain and have a
+            # unique extension to domain
+            base_valuation = domain.base_field().valuation(prime)
+            return self.create_key_and_extra_args_from_valuation(domain, base_valuation)
+
```

Comment:
We should have a different error message when the user inputs a prime ideal in the domain.